### PR TITLE
Add some Blob, File, FileList functions

### PR DESCRIPTION
**Pull request by zudov**
These become quite essential when you want to handle file upload.

### CODE REVIEW

**Comment by zudov** [comment]
Perhaps this should better go into a separate library.

If we would start adding bindings for all web APIs `purescript-dom` would grow into monstrosity. I like the idea of only keeping the basic type definitions in `purescript-dom`, providing bindings to methods and properties in other libraries.

**Comment by garyb** [comment]
Yeah, maybe, I don't think it's a problem now though.

File: src/DOM/File/FileList.purs [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+foreign import length :: FileList -> Int
+
+-- | Get `File` at the certain position
+foreign import itemImpl :: Int -> FileList -> Nullable File
+
+item :: Int -> FileList -> Maybe File
+item i l = toMaybe (itemImpl i l)
+
+toArray :: FileList -> Array File
+toArray fileList = Array.mapMaybe
```
Comment (garyb) [inline src/DOM/File/FileList.purs]: I'm not sure whether we should include this, I think it's the only function in the entire library that is not a type coercion/foreign read/direct mapping to an API method.

File: src/DOM/File/FileList.purs [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+import DOM.File.Types (FileList(), File())
+
+-- | Number of files in the `FileList` object.
+foreign import length :: FileList -> Int
+
+-- | Get `File` at the certain position
+foreign import itemImpl :: Int -> FileList -> Nullable File
+
+item :: Int -> FileList -> Maybe File
+item i l = toMaybe (itemImpl i l)
```
Comment (garyb) [inline src/DOM/File/FileList.purs]: Likewise this actually.

They're both definitely useful, but this library is supposed to be providing a low level unopinionated interface, so I think the `Nullable File` version is fine.

Comment (zudov) [inline src/DOM/File/FileList.purs]: Makes sense.

File: src/DOM/File/FileList.purs (hunk 1 above)
Comment (zudov) [inline src/DOM/File/FileList.purs]: Yeah, that's a bit too much.

File: src/DOM/File/Blob.purs [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+
+foreign import typeImpl :: Blob -> String
+
+-- | A string indicating the MIME type of the data contained in the `Blob`.
+-- | Returns `Nothing` if the type is unknown.
+type_ :: Blob -> Maybe String
+type_ blob = let blobType = typeImpl blob
+             in if blobType == ""
+                then Nothing
+                else Just blobType
```
Comment (zudov) [inline src/DOM/File/Blob.purs]: @garyb Is this under the question too?

Comment (garyb) [inline src/DOM/File/Blob.purs]: I stared at this one a while and wasn't too sure... but yeah, maybe we just expose the string even if it's empty. I guess ideally this should use a type other than string anyway, but `MimeType` just exists in `affjax` at the moment, I've not broken out that stuff yet.

Comment (zudov) [inline src/DOM/File/Blob.purs]: I also thought that it would be nice to have `MimeType` separated from `affjax`.
I could help with doing that.

Comment (garyb) [inline src/DOM/File/Blob.purs]: I've just started doing that myself actually :smile:

Comment (zudov) [inline src/DOM/File/Blob.purs]: Great!

Comment (garyb) [inline src/DOM/File/Blob.purs]: There are two existing mime packages it seems, but outdated/dodgy implementation. This one will be `purescript-media-types` and the type will be called `MediaType` as apparently that's the preferred terminology these days. (MIME is email specific).

Comment (zudov) [inline src/DOM/File/Blob.purs]: Sounds good

Comment (garyb) [inline src/DOM/File/Blob.purs]: Ok done, and published: https://github.com/purescript-contrib/purescript-media-types

v0.1.0 for now, as everything will need bumping when we do the core libraries for 0.8.

Comment (zudov) [inline src/DOM/File/Blob.purs]: Thanks!

I've made use of it in the https://github.com/purescript-contrib/purescript-dom/commit/f172a479c8af817c0f993fb3bb3ab3a6426245bf

**Comment by garyb** [comment]
Thanks! Last thing: there are some unused imports now that we're mostly relying on the FFI implementations.

**Comment by zudov** [comment]
Oh, yes, I'll clean them now.

**Comment by zudov** [comment]
Rebased. It should be ready to merge now.

File: src/DOM/File/File.purs [hunk 4]
```diff
@@ -0,0 +1,18 @@
...
+  , lastModifiedDate
+  ) where
+
+import DOM.File.Types (File(), Blob())
+
+import Unsafe.Coerce (unsafeCoerce)
+import Data.Date (JSDate())
+
+toBlob :: File -> Blob
+toBlob = unsafeCoerce
```
Comment (garyb) [inline src/DOM/File/File.purs]: Oh no... this is the last suggestion, I promise! :wink:

I think this should probably be `fileToBlob` and in the `Types` module instead, as the coercions were done that way for the other types in this library.

Comment (zudov) [inline src/DOM/File/File.purs]: Right. I'll try to stick to that scheme in the future. --amended the commit.

**Comment by garyb** [comment]
Thanks! Sorry about the multiple rounds of amends.

**Comment by zudov** [comment]
Amends are good! Thank you.

**Pull request merged**